### PR TITLE
fix(web): mnemonic keystrokes w FF keymapping

### LINKED ISSUE #1125: bug(web): Philippines SIL keyboard not compatible with Firefox
There is an issue with FF (problem is on FireFox in Windows and also Linux, not tested on OSX) is demonstrated with this keyboard:
[Philippines SIL](https://keymanweb.com/?_ga=2.178222438.2116510910.1534870017-227799419.1521670190#tl,Keyboard_sil_philippines)

This keyboard has various deadkeys which all seem to work except for problems with the underscore deadkey:

In FireFox, if you use your keyboard and type "_c" you get ```_c```. If you use the osk and your mouse and select ```_``` + `c` you get `ç` (as intended). In Chrome, you always get the cedilla whichever way you type it. Other keystrokes to try would be ```_o```, ```_a```, ```_w``` and ```_1```. Also, ```&:``` is supposed to output ```÷``` and it only works if you use the osk with mouse selection of those keys in Firefox. Again they work fine in Chrome.

Help file with keystrokes is here: https://help.keyman.com/keyboard/sil_philippines

### PR DESCRIPTION
**Pull request by jahorton**
Fixes #1125.

Looks like the browser-based keymapping was done too late during keystroke analysis.

Note that moving said keymapping to earlier within the same function may slightly affect a few `.vkCode` values in the legacy path.  That said... the only use I see for `.vkCode` in further processing is a `> 255` check, which would _not_ be affected.  It's also quite possible that if there are side-effects, they're good ones - given how niche the scenario is, I wouldn't be surprised if this actually fixed a similar bug on that legacy path that simply hasn't yet been discovered.

### CODE REVIEW

File: common/core/web/keyboard-processor/src/text/keyMapping.ts [hunk 1]
```diff
@@ -18,6 +18,7 @@ namespace com.keyman {
       //ffie['k107'] = 187; // =    // FF 3.0 // I2062
       this.FF['k61'] = 187;  // =   // FF 2.0
       this.FF['k59'] = 186;  // ;
+      this.FF['k173'] = 189; // -/_
```
Comment (mcdurdin) [inline common/core/web/keyboard-processor/src/text/keyMapping.ts:21]: This is a new mapping. Can you describe how you found that this was needed, and which versions of Firefox it has been needed for? Is this true for FF on all platforms?

Are there any other mappings we are missing? Also given those other commented entries from 2012, might be a good opportunity to clean them out.

File: web/source/dom/preProcessor.ts [hunk 2]
```diff
@@ -157,6 +157,19 @@ namespace com.keyman.dom {
        */
       s.Lmodifiers |= (e.metaKey ? modifierCodes['META']: 0);
 
+      // Physically-typed keys require use of a 'desktop' form factor and thus are based on a virtual "physical" Device.
+      s.device = keyman.util.physicalDevice.coreSpec;
+
+      // Check for any browser-based keymapping before proceeding further.
+      // This is needed _now_ for mnenomic keyboards.  (See https://github.com/keymanapp/keyman/issues/1125.)
+      if(!keyman.isEmbedded && s.device.browser == utils.Browser.Firefox) {
+        // I1466 - Convert the - keycode on mnemonic as well as positional layouts
```
Comment (mcdurdin) [inline web/source/dom/preProcessor.ts:166]: This comment is now outdated.

File: web/source/dom/preProcessor.ts [hunk 3]
```diff
@@ -157,6 +157,19 @@ namespace com.keyman.dom {
        */
       s.Lmodifiers |= (e.metaKey ? modifierCodes['META']: 0);
 
+      // Physically-typed keys require use of a 'desktop' form factor and thus are based on a virtual "physical" Device.
+      s.device = keyman.util.physicalDevice.coreSpec;
+
+      // Check for any browser-based keymapping before proceeding further.
+      // This is needed _now_ for mnenomic keyboards.  (See https://github.com/keymanapp/keyman/issues/1125.)
```
Comment (mcdurdin) [inline web/source/dom/preProcessor.ts:164]:
```suggestion
      // Browser key identifiers are not completely consistent; for example FireFox has 
      // several different key identifiers for US punctuation; see `Keymapping.browserMap` for 
      // details. So, check for any browser-specific key identifier remapping required before
      // applying mnemonic transforms.
      // See https://github.com/keymanapp/keyman/issues/1125
```

**Review by mcdurdin — CHANGES_REQUESTED** [review]
This LGTM but for some additional documentation. This problem I guess has been around for quite a while, so it would be good to be sure there are no more associated nasties lurking with it.

File: web/source/dom/preProcessor.ts (hunk 2 above)
Comment (jahorton) [reply web/source/dom/preProcessor.ts:166]: True, but oddly still very appropriate - that's literally one thing this PR is fixing.

The new comments (L163-164) are a more general version of that, so yeah, no need for the redundancy.  Will delete. :+1:

File: common/core/web/keyboard-processor/src/text/keyMapping.ts (hunk 1 above)
Comment (jahorton) [reply common/core/web/keyboard-processor/src/text/keyMapping.ts:21]: > This is a new mapping. Can you describe how you found that this was needed

Uh... I read the issue?  Noting that it was a FF-specific thing, my mind _very_ quickly went to the FF-specific keymapping section.  Then it was a matter of determining it was the `_` not being mapped, then confirming that the keycode didn't match Chrome's keycode.

For your other questions... they'll take additional time and effort to answer.  I'm fairly certain "true for FF on all platforms" == true, but the others... :shrug:

Comment (jahorton) [reply common/core/web/keyboard-processor/src/text/keyMapping.ts:21]: https://css-tricks.com/snippets/javascript/javascript-keycodes/ strongly suggests it's just these three.

https://www.freecodecamp.org/news/javascript-keycode-list-keypress-event-key-codes/ reports a few others that are irrelevant, as they're media control keys.

I haven't found any references that indicate any form of platform-specific behavior for these three keys.  The closest thing is a Mac-specific behavior... for other keys... for the `keypress` event only.  (Check the first link.)

Finally found the old MDN table, which could _probably_ use an update:  https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyCode

That said, note the details about Gecko (the Firefox engine).  A Google search shows that version 29 was released April 29th, 2014, so this has been around for a while.

Comment (mcdurdin) [reply common/core/web/keyboard-processor/src/text/keyMapping.ts:22]: The comments suggest Firefox 2.0 ... with some discrepancies fixed in 3.0. That's a long time ago. Like 2005 or so.

https://github.com/keymanapp/keyman/blob/v2.0.377/source/kmwkeymaps.js#L26-L31

**Review by mcdurdin — APPROVED** [review]

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 14.0.249-beta](https://keyman.com/downloads/releases/beta/14.0.249)

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 15.0.19-alpha](https://keyman.com/downloads/releases/alpha/15.0.19)

**Pull request merged**